controllers/controllers/clusters/vsphere.go

```diff
@@ -2,19 +2,153 @@ package clusters
 
 import (
 	"context"
+	"fmt"
+	"os"
+
+	"github.com/go-logr/logr"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/aws/eks-anywhere/controllers/controllers/reconciler"
+	"github.com/aws/eks-anywhere/controllers/controllers/resource"
 	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	c "github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/providers/vsphere"
+	releasev1alpha1 "github.com/aws/eks-anywhere/release/api/v1alpha1"
 )
 
+// Struct that holds common methods and properties
 type VSphereReconciler struct {
+	Client    client.Client
+	Log       logr.Logger
+	Validator *vsphere.Validator
+	Defaulter *vsphere.Defaulter
+}
+
+type VSphereClusterReconciler struct {
+	VSphereReconciler
 	*providerClusterReconciler
+
+	capiResourceFetcher *resource.CapiResourceFetcher
 }
 
-func NewVSphereReconciler() *VSphereReconciler {
-	return &VSphereReconciler{providerClusterReconciler: &providerClusterReconciler{}}
+func NewVSphereReconciler(client client.Client, log logr.Logger, validator *vsphere.Validator, defaulter *vsphere.Defaulter) *VSphereClusterReconciler {
+	capiResourceFetcher := resource.NewCAPIResourceFetcher(client, log)
+	return &VSphereClusterReconciler{
+		VSphereReconciler: VSphereReconciler{
+			Client:    client,
+			Log:       log,
+			Validator: validator,
+			Defaulter: defaulter,
+		},
+		providerClusterReconciler: &providerClusterReconciler{},
+		capiResourceFetcher:       capiResourceFetcher,
+	}
 }
 
-func (v *VSphereReconciler) Reconcile(ctx context.Context, cluster *anywherev1.Cluster) (reconciler.Result, error) {
+func (v *VSphereReconciler) VsphereCredentials(ctx context.Context) (*apiv1.Secret, error) {
+	secret := &apiv1.Secret{}
+	secretKey := client.ObjectKey{
+		Namespace: "eksa-system",
+		Name:      vsphere.CredentialsObjectName,
+	}
+	if err := v.Client.Get(ctx, secretKey, secret); err != nil {
+		return nil, err
+	}
+	return secret, nil
+}
+
+func (v *VSphereReconciler) SetupEnvsAndDefaults(ctx context.Context, vsphereDatacenter *anywherev1.VSphereDatacenterConfig) error {
```

Comment:
I like we are reusing this code for the multiple controllers ❤️ 

However, could we keep this as standalone methods (or maybe on a separate struct)?
It's ok for now, but I fear this might end up being another "vsphere provider" case, where if we add everything to one struct it would become to difficult to handle

Reply:
done